services/vision/mlvision/ml_model.go

```diff
@@ -0,0 +1,227 @@
+	ModelName string `json:"mlmodel_name"`
+}
+
+func registerMLModelVisionService(
+	ctx context.Context,
+	name string,
+	params *MLModelConfig,
+	r robot.Robot,
+	logger golog.Logger,
+) (vision.Service, error) {
+	_, span := trace.StartSpan(ctx, "service::vision::registerMLModelVisionService")
+	defer span.End()
+
+	mlm, err := mlmodel.FromRobot(r, params.ModelName)
+	if err != nil {
+		return nil, err
+	}
+	classifierFunc, err := attemptToBuildClassifier(mlm)
+	if err != nil {
+		logger.Infof("%v", errors.Wrapf(err, "was not able to turn ml model %q into a classifier", params.ModelName))
+	} else {
+		logger.Infof("model %q fulfills a vision service classifier", params.ModelName)
+	}
+	detectorFunc, err := attemptToBuildDetector(mlm)
+	if err != nil {
+		logger.Infof("%v", errors.Wrapf(err, "was not able to turn ml model %q into a detector", params.ModelName))
+	} else {
+		logger.Infof("model %q fulfills a vision service detector", params.ModelName)
+	}
+	segmenter3DFunc, err := attemptToBuild3DSegmenter(mlm)
+	if err != nil {
+		logger.Infof("%v", errors.Wrapf(err, "was not able to turn ml model %q into a 3D segmenter", params.ModelName))
+	} else {
+		logger.Infof("model %q fulfills a vision service 3D segmenter", params.ModelName)
+	}
+	// Don't return a close function, because you don't want to close the underlying ML service
+	return vision.NewService(name, r, nil, classifierFunc, detectorFunc, segmenter3DFunc)
+}
+
+// Unpack output based on expected type and force it into a []float64.
+func unpack(inMap map[string]interface{}, name string, md mlmodel.MLMetadata) []float64 {
+	var out []float64
+	me := inMap[name]
+	switch getTensorTypeFromName(name, md) {
+	case UInt8:
+		temp := me.([]uint8)
+		for _, t := range temp {
+			out = append(out, float64(t))
+		}
+	case Float32:
+		temp := me.([]float32)
+		for _, p := range temp {
+			out = append(out, float64(p))
+		}
+	default:
+		return nil
+	}
+	return out
+}
+
+// getTensorTypeFromName uses the metadata to find the expected type of the tensor.
+func getTensorTypeFromName(name string, md mlmodel.MLMetadata) string {
+	for _, o := range md.Outputs {
+		if strings.Contains(strings.ToLower(o.Name), strings.ToLower(name)) {
+			return o.DataType
+		}
+		if strings.Contains(name, "output") {
+			_, after, _ := strings.Cut(name, "output")
+			saveI, err := strconv.Atoi(after)
+			if err != nil {
+				return ""
+			}
+			return md.Outputs[saveI].DataType
+		}
+	}
+	return ""
+}
+
+// getLabelsFromMetadata returns a slice of strings--the intended labels.
+func getLabelsFromMetadata(md mlmodel.MLMetadata) []string {
+	for _, o := range md.Outputs {
+		if strings.Contains(o.Name, "category") || strings.Contains(o.Name, "probability") {
+			if labelPath, ok := o.Extra["labels"]; ok {
+				labels := []string{}
+				f, err := os.Open(labelPath.(string))
+				if err != nil {
+					return nil
+				}
+				defer func() {
+					if err := f.Close(); err != nil {
+						panic(err)
```

Comment:
our rule is to never panic unless absolutely necessary. If this is not properly caught by the caller (and their callers), it will crash the whole process.